src/ra_server.erl

```diff
@@ -3314,6 +3443,186 @@ after_log_append_reply(Cmd, Idx, Term, Effects0) ->
+                      PeerMacVer < MacVer;
+                  _ ->
+                      %% We don't have any details about the peer, we ask.
+                      true
+              end,
+    case SendRpc of
+        true ->
+            %% We ask for all info keys currently. If we ask for specific
+            %% keys, we will have to handle merging of already known info keys
+            %% and updates.
+            Id = id(State),
+            Command = #info_rpc{from = Id,
+                                term = CurTerm,
+                                keys = [machine_version]},
+            [{send_rpc, PeerId, Command}];
+        false ->
+            []
+    end.
+
+info_reply_effect(
+  #{current_term := CurTerm} = State,
+  #info_rpc{from = FromId, keys = Keys}) ->
+    Id = id(State),
+    Info = ra_server_info(State, Keys),
+    InfoReply = #info_reply{from = Id,
+                            term = CurTerm,
+                            keys = Keys,
+                            info = Info},
+    {cast, FromId, InfoReply}.
+
+empty_info_reply_effect(
+  #{current_term := CurTerm} = State,
+  #info_rpc{from = FromId, keys = Keys}) ->
+    Id = id(State),
+    InfoReply = #info_reply{from = Id,
+                            term = CurTerm,
+                            keys = Keys,
+                            info = undefined},
+    {cast, FromId, InfoReply}.
+
+ra_server_info(State) ->
+    MacVer = machine_version(State),
+    #{machine_version => MacVer}.
+
+ra_server_info(State, Keys) ->
+    %% Note that the node that asked may ask for keys we don't know.
+    Info0 = ra_server_info(State),
+    Info1 = maps:filter(
+              fun(Key, _Value) ->
+                      lists:member(Key, Keys)
+              end, Info0),
+    Info1.
+
+handle_info_reply(
+  #{cluster := Cluster} = State,
+  #info_reply{from = PeerId, keys = Keys, info = Info}) ->
+    PeerState0 = maps:get(PeerId, Cluster),
+    PeerState1 = case Info of
+                     undefined ->
+                         %% The peer is in a Raft state where it can't give
+                         %% info yet.
+                         lists:foldl(
+                           fun(Key, PS) ->
+                                   maps:remove(Key, PS)
+                           end, PeerState0, Keys);
+                     _ ->
+                         lists:foldl(
+                           fun(Key, PS) ->
+                                   Value = maps:get(Key, Info),
+                                   PS#{Key => Value}
+                           end, PeerState0, Keys)
+                 end,
+    Cluster1 = Cluster#{PeerId => PeerState1},
+    State1 = State#{cluster => Cluster1},
+    determine_if_machine_upgrade_allowed(State1).
+
+determine_if_machine_upgrade_allowed(
+  #{cfg := #cfg{effective_machine_version = EffectiveMacVer,
+                log_id = LogId}} = State) ->
+    Effects = case has_enough_peer_info(State) of
```

Comment:
do we need `has_enough_peer_info` at all if the later `get_max_supported_machine_version` could handle the case where some peers do not yet have a common machine version?

Reply:
I merged the two functions.